mlflow/projects/__init__.py

```diff
@@ -44,7 +44,12 @@ def _run(uri, entry_point="main", version=None, parameters=None, experiment_id=N
     if run_id:
         active_run = tracking.get_service().get_run(run_id)
     else:
-        active_run = _create_run(uri, exp_id, work_dir, entry_point, parameters)
+        active_run = _create_run(uri, exp_id, work_dir, entry_point)
+
+    # log all params
```

Comment:
Small nit - if we log the parameters here, we'll end up logging them twice for Databricks or asynchronous local runs: once when we create the run on the client and a second time when we invoke `mlflow run` on our Databricks cluster or locally (the latter happens for async local runs). Logging twice is ok since the params don't change, but we could push the log-all-params logic into the `_create_run` branch above to only log params at run creation time.

Reply:
I was thinking about that and I chose to add this here for the following reasons.

- This was the common point that all paths for MLflow runs go through as soon as the `run_uuid` is available
- If reusing a `run_uuid` (for instance though projects or after a training run crashes and restarting from N-th epoch checkpoint), this would serve as a common point to ensure that the same parameters are used, as in if a previously logged param value is different, `mlflow run` command will throw right here.

Reply:
I am not sure about using params as a correctness check. If we want to have an extra test that the run is what we expect it to be (e.g.  maybe there was a rogue environment variable), we should add code dedicated for that. It would still be possible to log a run to a different run Id by mistake if the parameters matched. 

I don't think the double logging matters too much, but I have a slight preference for moving the code to _create_run to avoid double logging and also I thinks it's a a bit cleaner design.

Reply:
I am not suggesting we use params as a correctness check, but my concern here is that we have a in and out branches in this code. Incoming as an already created run and reusing `run_uuid` or creating a new one. Outgoing branches as cloud vs local (sync vs async). Some might log, some dont. So it defeats the purpose of MLflow project logging any params--we can completely leave it to the actual code to take care of that.

The other proposal is we log only in the create run path and if re-using run_uuid: we query tracking server to query all logged params and ensure correctness. Not sure if that is worth it.